firmware/library/L1_Drivers/ssp.hpp

```diff
@@ -70,3 +72,215 @@ class SspInterface
+                (master_mode_ << kMasterModeBit);
+            peripheral_initialized = 1;
+        }
+        return peripheral_initialized;
+    }
+
+    // Renamed from get_status()
+    // Checks if the SSP controller is idle.
+    // @return 1 - the controller is sending or receiving a data frame.
+    // @return 0 - the controller is idle.
+    bool GetTransferStatus() override
+    {
+        return (ssp_registers[static_cast<uint32_t>(pssp_)]->SR &
+            (1 << kDataLineIdleBit));
+    }
+
+    // Transfers a data frame to an external device using the SSP
+    // data register. This functions for both transmitting and
+    // receiving data. It is recommended this region be protected
+    // by a mutex.
+    // @param data - information to be placed in data register
+    // @return - received data from external device
+    uint16_t Transfer(uint16_t data) override
+    {
+        ssp_registers[static_cast<uint32_t>(pssp_)]->DR = data;
+        while (GetTransferStatus()) { continue; }
+        return static_cast<uint16_t>(
+                ssp_registers[static_cast<uint32_t>(pssp_)]->DR);
+    }
+
+    // Set the PCONP bits to power on peripheral
+    // uses the pssp_ set in constructor
+    void SetPeripheralPower() override
+    {
+        pconp_bits_ =
+            (pssp_ == Peripheral::kSsp0) ? kPconp0 : kPconp1;
+    }
+
+    // Sets the various modes for the Peripheral
+    // @param mode - master or slave mode
+    // @param frame - format for Peripheral data to use
+    // @param size - number of bits per frame
+    void SetPeripheralMode(
+        MasterSlaveMode mode, FrameMode frame, DataSize size) override
+    {
+        master_mode_ = mode;
+        frame_format_ = frame;
+        data_size_ = size;
+    }
+
+    // Sets the clock rate for the Peripheral
+    void SetClock(bool polarity, bool phase,
+        uint8_t divider, uint8_t prescaler) override
+    {
+        clock_polarity_ = polarity;
+        clock_phase_ = phase;
+        clock_divider_ = divider;
+        clock_prescaler_ = prescaler;
+    }
+
+    ~Ssp()
+    {
+        // do nothing
+    }
+
+ private:
+    PinConfigureInterface & mosi_;
+    PinConfigureInterface & miso_;
+    PinConfigureInterface & sck_;
+
+    PinConfigure mosi_pin_;
+    PinConfigure miso_pin_;
+    PinConfigure sck_pin_;
+
+    // SSP member variables
+    Peripheral pssp_;                // SSP interfaces
+    PowerOn pconp_bits_;
+    MasterSlaveMode master_mode_;
+    DataSize data_size_;             // data size, number of bits
+    FrameMode frame_format_;         // frame format
+    bool clock_polarity_;            // clock polarity
+    bool clock_phase_;               // clock phase
+    // clock_divider_ is the number of prescaler-output clocks per bit
+    // see page 611 of UM10562 for SCR-Serial Clock Rate for more info
+    uint8_t clock_divider_;
+    uint8_t clock_prescaler_;        // clock prescaler

```

Comment:
Ok, you may hate me for doing this, but rather than holding these variables within the class, you can save memory by simply use the registers. 

You should create getters and setters for each of these modes, by reading from the registers. This is also helpful, because the registers are the ground truth for the values. There is potential for your class to be out of sync with the registers.

The conclusion, is don't store these values in class variables, if you can derive their values from the registers.

Reply:
This approach would essentially break apart the Initialize() function wouldn't it? Since it is no longer using these variables to initialize everything at once? 

I think I will also have to redo each constructor as well? For example, the default constructor sets everything up to be SPI master when Initialize() runs. If I remove all these variables, it won't have the default ability anymore...There are a few things I can automatically set given the peripheral option (SSP0 vs SSP1) and just make Initialize() the power up function... Hmm, let me see what I can do.

Reply:
You correct. The idea here, would be to have Initialize simply enable the SPI and power it on. Or maybe having SPI initialize sets up some defaults and then you add an enable method to power on the SPI. If initialize was used just to power everything on, the code one would write would be the same. You would set all of the configurations using the appropriate methods, and call initialize at the end to enable the spi device. You could also introduce a `SetupDefaults` method that can be used on its own, or maybe even in the init function if none of the other methods have been called. Tell me what you think and what you think would be a cleaner design.

One thing, I don't think we should do is use internal variables for the class.

@SameerAzer What do you think?

Reply:
Would it be acceptable to group like items together, similar to my current setter functions, or would it be better to make an individual one for each?